### PR TITLE
refactor(anchor-navigation): convert to typescript

### PR DESCRIPTION
**Pull request by Parsium**
### Proposed behaviour

- Convert to `.component` files to typescript
- Convert jest spec file to typescript
- Convert `.style` files to typescript
- Remove `d.ts` files
- Convert index file to typescript and export types for consumer use
- Convert test story file from `.mdx` to `.tsx`
    - Required due to an issue where Storybook wraps component in a `MDXCreateElement` wrapper, which causes the children type checks in `AnchorNavigation.component.tsx` to fail as the `displayName` is recognised as `MDXCreateElement`. This seems to be a problem when the template bind pattern is used for stories (see [this comment](https://github.com/storybookjs/storybook/issues/11542#issuecomment-665193303) for details).
- Add `@types/react-is` to dev dependencies - required for checking element is a React Fragment
- Change type of `target` prop of `AnchorNavigationItem` from `React.Ref<HTMLElement> | undefined` to `React.RefObject<HTMLElement | undefined>`.

### Current behaviour

- Component is entirely written in javascript

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

Double check for any possible regressions introduced from this refactor.

### CODE REVIEW

File: src/components/anchor-navigation/anchor-navigation.spec.tsx [hunk 1]
```diff
@@ -144,26 +147,29 @@ describe("AnchorNavigation", () => {
       });
       wrapper.update();
       expectNavigationItemToBeSelected(index, wrapper);
-
-      if (hasFocusableElement) {
```
Comment (Parsium) [inline src/components/anchor-navigation/anchor-navigation.spec.tsx:148]: Split this test into two, one for testing the click and scroll to section, and the other click and focus on sections with focusable content, since eslint does not recommend conditional expect statements.

File: src/components/anchor-navigation/anchor-navigation.spec.tsx [hunk 2]
```diff
@@ -173,15 +179,28 @@ describe("AnchorNavigation", () => {
           jest.advanceTimersByTime(15);
         });
 
-        if (hasFocusableElement) {
```
Comment (Parsium) [inline src/components/anchor-navigation/anchor-navigation.spec.tsx:176]: Likewise here

File: src/components/anchor-navigation/anchor-navigation.component.tsx [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  const navigationRef = useRef<HTMLUListElement>(null);
+
+  const isUserScroll = useRef(true);
+
+  const isUserScrollTimer = useRef<NodeJS.Timeout>();
+
+  const setSelectedAnchorBasedOnScroll = useCallback(() => {
+    // istanbul ignore if
+    // function is called only after component is rendered, so ref cannot hold a null value
+    if (navigationRef.current === null) return;
```
Comment (edleeks87) [inline src/components/anchor-navigation/anchor-navigation.component.tsx:97]: you could probably wrap the whole function in `if (navigationRef.current !== null) { ...}` and ignore else here, again not critical

File: src/components/anchor-navigation/anchor-navigation.spec.tsx [hunk 4]
```diff
@@ -243,23 +262,28 @@ describe("AnchorNavigation", () => {
   ])(
     "scroll triggers selection of proper navigation item based on the navigation and sections top position",
     (scrollPosition, selectedAnchorIndex) => {
+      const SECTION_VISIBILITY_OFFSET = 200;
```
Comment (edleeks87) [inline src/components/anchor-navigation/anchor-navigation.spec.tsx:265]: If we move these to the top of the file it will avoid initialising them on each loop

Comment (edleeks87) [inline src/components/anchor-navigation/anchor-navigation.spec.tsx:265]: If we move these to the top of the file it will avoid initialising them on each loop

File: src/components/anchor-navigation/anchor-navigation.component.tsx [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+      firstFocusableElement.focus({ preventScroll: true });
+    }
+  };
+
+  const scrollToSection = (index: number): void => {
+    const sectionToScroll = sectionRefs.current[index].current;
+
+    // istanbul ignore if
+    // function is called only after component is rendered, so ref cannot hold a null value
+    if (sectionToScroll === null) return;
```
[the remark this answers is not in the document]
Comment (Parsium) [reply src/components/anchor-navigation/anchor-navigation.component.tsx:165]: Personally I'd prefer keeping this in for readability, as we're making the return behaviour for this edge case explicit :)

**Review by edleeks87 — APPROVED** [review]

File: src/components/anchor-navigation/anchor-navigation-test.stories.tsx [hunk 6]
```diff
@@ -77,14 +83,17 @@ export const AnchorNavigationStory = () => {
     </AnchorNavigation>
   );
 };
+DefaultStory.storyName = "default";
 
-export const AnchorNavigationInFullScreenDialogStory = () => {
+export const InFullScreenDialogStory: ComponentStory<
```
Comment (DipperTheDan) [inline src/components/anchor-navigation/anchor-navigation-test.stories.tsx:88]: Something we need to consider, maybe not in this PR, but is to fix this: 
Screenshot 2022-06-08 at 14 07 02

I think there are parameters in Storybook to specify that the code should be shown and not that ⬆️

**Review by DipperTheDan — APPROVED** [review]

**Review by ZhuoyuJin — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]
